### ISSUE TITLE
Illegal Memory Access with Mixed Precision

### ISSUE THREAD

**Issue by HP-eter** [microsoft/aurora#121]
Hi, I'm running into a blocking issue when fine-tuning **Aurora** on a 95 GB H100 GPU using **Mixed Precision Training** (either via `bf16_mode` or `torch.autocast`).

---

### Setup

- **GPU:** NVIDIA H100 (95 GB)
- **Precision Mode:** BF16 (via `bf16_mode`) and/or `torch.autocast`
- **Batch Size:** 1 (with gradient accumulation over 4 steps)
- **Model:** Aurora
- **Fine-tuning Target:** 9 new **surface** variables  
- **Input to Model:** All pre-trained surface, static, and atmospheric variables  
- **Loss Function:** The Aurora loss as described in the original paper

---

###  What’s happening?

- The forward pass (`model(batch_in)`) runs **without any problems**.
- The error occurs **at `loss.backward()`**, when Mixed Precision is enabled.
- Specifically, I get an **illegal memory access** error that causes training to fail.

This only happens when Mixed Precision is active.

---

### What I’ve tried

- **Disabled BF16 & Autocast →**
  - ✅ No illegal memory access  
  - ❌ But eventually get an **out-of-memory (OOM)** error after 4–5 steps (full precision consumes about 90 GB during the first couple steps, and > 95 GB later)
  
- **Disabled Gradient Checkpointing (while keeping BF16 or Autocast enabled) →**
  - ✅ No illegal memory access  
  - ❌ But not viable because memory usage is too high without checkpointing

So it seems like the issue **only happens when Mixed Precision and Gradient Checkpointing are used together** ?
I could really use some help for debugging this.

**Comment by wesselb** [comment]
Hey all,

Thanks for opening an issue! Unfortunately this is a known problem. :( `bf16_mode` is experimental, and it might run into issues like this.

Since you're using a 95 GB GPU, could you wrap the forward in `torch.autocast("cuda", dtype=torch.blofat16)`, turn off `bf16_mode`, and turn on gradient checkpointing?

If you get illegal memory access errors, you might want to try adjusting the checkpointing strategy [here](https://github.com/microsoft/aurora/blob/main/aurora/model/aurora.py#L515). E.g., try this:

``` python
        module_names = (
            # "Perceiver3DEncoder",
            # "Swin3DTransformerBackbone",
            "Basic3DEncoderLayer",
            "Basic3DDecoderLayer",
            # "Perceiver3DDecoder",
            # "LinearPatchReconstruction",
        )
```

Hope you can make it work!
